src/main/java/com/liferay/faces/util/cache/internal/ConcurrentCacheMaxCapacityLRUImpl.java

```diff
@@ -0,0 +1,218 @@
+	 */
+	private void removeLeastRecentlyUsedCacheValuesIfNecessary(K key) {
+
+		// Don't synchronize on the ConcurrentHashMap in case it synchronizes on itself internally (avoid locking on
+		// reads).
+		synchronized (maxCapacity) {
+
+			if ((internalCache.size() >= maxCapacity) && !internalCache.containsKey(key)) {
+
+				Set<Map.Entry<K, CachedValue<V>>> entrySet = internalCache.entrySet();
+				Map.Entry<K, CachedValue<V>> leastRecentlyAccessedEntry = null;
+
+				for (Map.Entry<K, CachedValue<V>> entry : entrySet) {
+
+					if (leastRecentlyAccessedEntry != null) {
+
+						CachedValue<V> cachedValue = entry.getValue();
+						CachedValue<V> leastRecentlyAccessedCacheValue = leastRecentlyAccessedEntry.getValue();
+
+						if (cachedValue.wasLastAccessedBefore(leastRecentlyAccessedCacheValue)) {
+							leastRecentlyAccessedEntry = entry;
+						}
+					}
+					else {
+						leastRecentlyAccessedEntry = entry;
+					}
+				}
+
+				if (leastRecentlyAccessedEntry != null) {
+					internalCache.remove(leastRecentlyAccessedEntry.getKey());
+				}
+			}
+		}
+	}
+
+	/**
+	 * This class tracks the last time a cached value was accessed in order to allow {@link
+	 * #removeLeastRecentlyUsedCacheValuesIfNecessary(java.lang.Object)} to compare access times of values. The last
+	 * access time is updated on CacheValue creation and {@link #getValue()} using {@link System#nanoTime()}. The access
+	 * time is stored inside a <a
+	 * href="https://stackoverflow.com/questions/3038203/is-there-any-point-in-using-a-volatile-long">
+	 * <code>volatile</code> <code>long</code></a> to ensure that the value of the <code>long</code> is always read and
+	 * written atomically. Although the access time is guaranteed to be read and written to atomically, it is not always
+	 * guaranteed to be perfectly up to date. For example, if one thread is checking the last access time (in order to
+	 * remove the least recently accessed value) and another happens to update it, the update may not be detected before
+	 * value is removed. That situation should rarely occur, and the negative impact of removing a recently used value
+	 * occasionally will likely impact performance much less than synchronizing updates to the access time value.
+	 */
+	private static final class CachedValue<V> {
+
+		// Private Final Data Members
+		private final V value;
+
+		//J-
+		// Private Volatile Data Members
+		//J+
+
+		// Reads and writes to volatile long primitives are atmoic:
+		// https://stackoverflow.com/questions/3038203/is-there-any-point-in-using-a-volatile-long
+		private volatile long lastAccessed;
+
+		public CachedValue(V value) {
+
+			this.lastAccessed = System.nanoTime();
+			this.value = value;
+		}
+
+		public long getLastAccessedNanoTime() {
+			return lastAccessed;
+		}
+
+		public V getValue() {
+
+			lastAccessed = System.nanoTime();
+
+			return value;
+		}
+
+		public boolean wasLastAccessedBefore(CachedValue otherCachedValue) {
```

Comment:
@stiemannkj1 Recommend changing the method name to "wasLessRecentlyAccessedBefore"